pkg/asset/cluster/aws/aws.go

```diff
@@ -29,48 +30,100 @@ func Metadata(clusterID, infraID string, config *types.InstallConfig) *awstypes.
 // PreTerraform performs any infrastructure initialization which must
 // happen before Terraform creates the remaining infrastructure.
 func PreTerraform(ctx context.Context, clusterID string, installConfig *installconfig.InstallConfig) error {
-	if len(installConfig.Config.Platform.AWS.Subnets) == 0 {
-		return nil
+	if len(installConfig.Config.Platform.AWS.Subnets) != 0 {
+
+		privateSubnets, err := installConfig.AWS.PrivateSubnets(ctx)
+		if err != nil {
+			return err
+		}
+
+		publicSubnets, err := installConfig.AWS.PublicSubnets(ctx)
+		if err != nil {
+			return err
+		}
+
+		//arns := make([]*string, 0, len(privateSubnets)+len(publicSubnets))
+		ids := make([]*string, 0, len(privateSubnets)+len(publicSubnets))
+		for id := range privateSubnets {
+			ids = append(ids, aws.String(id))
+		}
+		for id := range publicSubnets {
+			ids = append(ids, aws.String(id))
+		}
+
+		if err := tagSubnetEC2Instances(ctx, clusterID, installConfig, ids); err != nil {
+			return err
+		}
 	}
 
-	privateSubnets, err := installConfig.AWS.PrivateSubnets(ctx)
+	workerMachinePool, err := installConfig.Config.WorkerMachinePool()
 	if err != nil {
 		return err
 	}
 
-	publicSubnets, err := installConfig.AWS.PublicSubnets(ctx)
-	if err != nil {
-		return err
+	iamRoleNames := make([]*string, 0)
+
+	if installConfig.Config.ControlPlane.Platform.AWS.IAMRoleName != "" {
+		iamRoleNames = append(iamRoleNames, &installConfig.Config.ControlPlane.Platform.AWS.IAMRoleName)
 	}
 
-	//arns := make([]*string, 0, len(privateSubnets)+len(publicSubnets))
-	ids := make([]*string, 0, len(privateSubnets)+len(publicSubnets))
-	for id := range privateSubnets {
-		ids = append(ids, aws.String(id))
+	if workerMachinePool.Platform.AWS.IAMRoleName != "" {
+		iamRoleNames = append(iamRoleNames, &workerMachinePool.Platform.AWS.IAMRoleName)
 	}
-	for id := range publicSubnets {
-		ids = append(ids, aws.String(id))
+
+	if len(iamRoleNames) > 0 {
+		if err := tagIamRoles(ctx, clusterID, installConfig, iamRoleNames); err != nil {
+			return err
+		}
 	}
 
+	return nil
+}
+
+func tagSubnetEC2Instances(ctx context.Context, clusterID string, installConfig *installconfig.InstallConfig, instanceIds []*string) error {
 	session, err := installConfig.AWS.Session(ctx)
 	if err != nil {
 		return err
 	}
-
-	tags := []*ec2.Tag{
+	ec2Tags := []*ec2.Tag{
 		{
 			Key:   aws.String(fmt.Sprintf("kubernetes.io/cluster/%s", clusterID)),
 			Value: aws.String("shared"),
 		},
 	}
```

Comment:
Let's use a function to get the tag key and value rather than duplicate the logic for the ec2 tags and the iam tags.
```suggestion
	key, value := sharedTag(clusterID)
	ec2Tags := []*ec2.Tag{{Key: &key, Value: &value}}
```